table/server/underdb/hive/src/main/java/alluxio/table/under/hive/util/HMSShim.java

```diff
@@ -0,0 +1,172 @@
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.table.under.hive.util;
+
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.RetryingMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore;
+import org.apache.thrift.TApplicationException;
+import org.apache.thrift.TBase;
+import org.apache.thrift.TException;
+import org.apache.thrift.TServiceClient;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+/**
+ * Implements a shim layer for hive metastore client.
+ **/
+public class HMSShim implements HiveCompatibility {
+  private final TServiceClient mClient;
+
+  /**
+   * Constructor for HMSShim.
+   *
+   * @param client another client as delegate
+   */
+  public HMSShim(IMetaStoreClient client) {
+    // Because RetryingMetaStoreClient is itself a proxy, we need to get to the base class
+    while (Proxy.isProxyClass(client.getClass())) {
+      InvocationHandler handler = Proxy.getInvocationHandler(client);
+      if (handler.getClass().isAssignableFrom(RetryingMetaStoreClient.class)) {
+        client = getField(handler, "base");
+        continue;
+      }
+      // Other handlers can be added here
+      throw new RuntimeException("Unknown proxy handler for IMetaStoreClient");
+    }
+    mClient = getField(client, "client");
+  }
+
+  private static <T> T getField(Object object, String fieldName) {
+    try {
+      Field field = object.getClass().getDeclaredField(fieldName);
+      T result = null;
+      if (field.isAccessible()) {
+        result = (T) field.get(object);
+      } else {
+        field.setAccessible(true);
+        result = (T) field.get(object);
+        field.setAccessible(false);
+      }
+      return result;
+    } catch (SecurityException | NoSuchFieldException
+        | IllegalArgumentException | IllegalAccessException e) {
+      throw new RuntimeException("Unable to access the client through reflection", e);
+    }
+  }
+
+  private static Table deepCopy(Table table) {
+    return table.deepCopy();
+  }
+
+  private void sendBase(String methodName, TBase<?, ?> args) throws TException {
+    try {
+      Method sendBase = TServiceClient.class.getDeclaredMethod(
```

Comment:
Should this be saved/cached?

Reply:
i don't think there is a lot of value in caching these objects, mainly because the cost of reflection is going to be minimal compared to the network cost of these thrift calls.